Cleanups: make a gbfreadbuf, move units from defs and into dedicated file, reduce xstrdup,xmalloc use, reduce CString/QString roundtrips, more.

robertlipe opened the pull request:
I'd meant to have this in multiple submits, but here we are. Most of this is fairly mechanical. The KML change is worth a callout.

Our KML coordinate reader has been around the block. At various times,
it's been SAX, DOM, raw expat, expat front-ended for sane buffer management,
and is now a weird combination of C-level scanf and QString. I would have
guessed it to be faster than making two additional copies, the second of
which is parsed again after the first parse knocks down whitespace to make
the second level easy to parse for commas.

I'd have been very wrong.




With the existing code, parsing an arbitrary file of about 38,000
points takes about 9.8 seconds as measured by
$ time ./gpsbabel -t -i kml -f test.kml

With the new code, it's about .45 seconds.

It still passes validation. A 21x speedup that's also way simpler to read
isn't bad.

GPSBabelDeveloper committed:
In Teletype, seek over unknown instead of allocating, reading, and tossing.

GPSBabelDeveloper committed:
Make tpo less weird. Rely on xmalloc/xfree way less. Reduce internal copies.

GPSBabelDeveloper committed:
Add gbfreadbuf to simplify object storage.

GPSBabelDeveloper committed:
Make GPX internal promotions/reductions more specific.
Automated cleanups courtesy Clion.

GPSBabelDeveloper committed:
Lowranceusr: internal tightening of types, better scope management.

GPSBabelDeveloper committed:
Make WAYPT_SET act more like a function after macro expansion.

GPSBabelDeveloper committed:
Rework KML linestring coordinate reader. About a 21x speedup on a file
with 38,000 coordinates tuples.

GPSBabelDeveloper committed:
Largely automated cleanups in NMEA.

GPSBabelDeveloper committed:
Warning tweaks in defs for safety in macro expansion.

GPSBabelDeveloper committed:
Fix ordering in optional altitude test in KML reader

GPSBabelDeveloper committed:
A bunch of busy-work in coce that approximarly never is actually run but solely for the hope of hushing clang, clion, and codacy.

tsteven4 commented on kml.cc:
```diff
@@ -406,34 +406,28 @@ void wpt_icon(xg_string args, const QXmlStreamAttributes*)
-      trkpt->altitude = alt;
-    }
+  auto vecs = args.simplified().split(' ');
+  for(const auto& vec : vecs) {
```
I recommend running clazy when you touch any Qt container, especially in a range based for loop:
qmake QMAKE_CXX=/path/to/clazy; make
You may want to export a modified CLAZY_CHECKS variable, but clazy-range-loop is probably enabled by default as it is included in the level 1 checks.

> kml.cc:416:3: warning: c++11 range-loop might detach Qt container (QStringList) [-Wclazy-range-loop]
  for(const auto& vec : vecs) {
  ^
                        qAsConst( )

The clazy suggested fix will work, but an alternative I prefer when possible is to make the container const:
`const auto vecs = args.simplified.split(' ');`

robertlipe replied:
Integrated. Thanx. It drives me crazy that 'auto' doesn't always return the "best" type.

Should we work on a clazy pass as a presubmit test?

tsteven4 commented on kml.cc:
```diff
@@ -406,34 +406,28 @@ void wpt_icon(xg_string args, const QXmlStreamAttributes*)
-    }
+  auto vecs = args.simplified().split(' ');
+  for(const auto& vec : vecs) {
+    QStringList coords = vec.split(',');
```
I recommend making Qt containers const when possible as it makes it clear you don't want to modify the container and  it prevents issues with detach.

robertlipe replied:
Fixed as above.

tsteven4 commented on kml.cc:
```diff
@@ -406,34 +406,28 @@ void wpt_icon(xg_string args, const QXmlStreamAttributes*)
+    auto trkpt = new Waypoint;
 
+    if (csize == 3) {
+      trkpt->altitude = coords[2].toDouble();
+    }
+    if (csize == 2 || csize == 3) {
+      trkpt->latitude = coords[1].toDouble();
+      trkpt->longitude = coords[0].toDouble();
```
For read access I prefer at() over operator[] because it will never cause a detach, .i.e.
`coords.at(n)`
etc.
See the discussion at https://doc.qt.io/qt-5/qlist.html#details
However if the list is const then they are equivalent.

robertlipe replied:
They're now const.

GPSBabelDeveloper committed:
Mechanically move protos and structs for units.cc from defs.h to units.h

GPSBabelDeveloper committed:
Add util.h to various build files.

GPSBabelDeveloper committed:
Cleanup train keeps rollin' along
defs.h: use c++ true and false. Replace macros with constexprs.
gbser_win.cc don't use Windows style TRUE/FALSE.
kml.cc: integration suggestions.
msvc build: use units.h
